audit/base/base.go

```diff
@@ -0,0 +1,556 @@
+	for {
+		if len(c.sendCh) == 0 {
+			return
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+}
+
+// close handles closing the connection to the audit server. This is used by Close().
+// It ensures that the connection is only closed once.
+func (c *Client) close(ctx context.Context, stopSender bool) error {
+	c.closeOnceMu.Lock()
+	defer c.closeOnceMu.Unlock()
+
+	c.closeOnce.Do(func() {
+		if stopSender {
+			// This might be new to some of you, but you can send a channel over a channel.
+			// In this case, I create a channel and send it on the stopSend channel. When the
+			// sender receives the channel, it will close the channel I sent, and then we know that
+			// the sender is done.
+			sig := make(chan struct{}, 1)
+			c.stopSend <- sig
+			<-sig
+		}
+
+		if err := c.getErr(); err != nil {
+			ptr := c.conn.Load()
+			if ptr != nil {
+				(*ptr).CloseSend(ctx) // Ignore any error
+			}
+			return
+		}
+
+		ptr := c.conn.Load()
+		if ptr != nil {
+			if err := (*ptr).CloseSend(ctx); err != nil {
+				c.setErr(Error{Err: err, Category: ErrConnection})
+			}
+		}
+	})
+	return c.getErr()
+}
+
+// sender is the async sender for the audit client.
+func (c *Client) sender() {
+	defer close(c.done)
+
+	var ticker *time.Ticker
+	defer func() {
+		if ticker != nil {
+			ticker.Stop()
+		}
+	}()
+
+	var tickerCh <-chan time.Time
+
+	for {
+		if c.successSend && ticker == nil {
+			c.write(ctxBack, c.heartbeat)
+			ticker = time.NewTicker(c.heartbeatInterval)
+			tickerCh = ticker.C
+		}
+
+		select {
+		case sig := <-c.stopSend:
+			// Let the other side know we are done.
+			sig <- struct{}{}
+			return
+		case sm := <-c.sendCh:
+			c.write(sm.Data.Ctx, sm.Data.Msg)
+		case <-tickerCh:
+			c.write(ctxBack, c.heartbeat)
+		}
+	}
+}
+
+func (c *Client) write(ctx context.Context, msg msgs.Msg) {
+	ptr := c.conn.Load()
+	if ptr == nil {
+		return
+	}
+
+	if err := (*ptr).Write(ctx, msg); err != nil {
+		// Requeue the message if we can.
+		select {
+		case c.sendCh <- streamData[SendMsg]{Data: SendMsg{Ctx: ctx, Msg: msg}}:
+		default:
+			c.log.Error(fmt.Sprintf("audit message dropped due to queue being full: %v", err))
+		}
+
+		if err == context.Canceled {
+			c.log.Error(fmt.Sprintf("audit message had context cancellation: %s", err))
+			return
+		}
+		c.setErr(Error{Err: err, Category: ErrConnection})
+		return
+	}
+	c.successSend = true
+}
+
+// setErr sets the error for the client if the error is not already set. If the error being set it nil,
+// it will be set to nil.
+func (c *Client) setErr(err error) {
+	c.setErrMu.Lock()
+	defer c.setErrMu.Unlock()
+
+	if err == nil {
+		c.err.Store(nil)
+	}
+
+	if c.err.Load() != nil {
+		return
+	}
+
+	c.err.Store(&err)
+}
+
+// getErr gets the error for the client.
+func (c *Client) getErr() error {
+	if ptr := c.err.Load(); ptr != nil {
+		return *ptr
+	}
+	return nil
+}
+
+// streamData is a generic type for sending data to a channel.
+type streamData[T any] struct {
+	Data T
+	Err  error
```

Comment:
Do you ever actually use this `Err`?

Reply:
No, this is me just defaulting to a standard pattern I use. Almost always I'm communicating in a 1:1 message type even if async (with a promise). But in this one I'm not, so I removed this wrapper.